eng/doc/fips/README.md

```diff
@@ -209,31 +215,30 @@ The following sections describe how to enable FIPS mode and the effect of the `G
   - The Linux Kernel's FIPS mode sets the content of `/proc/sys/crypto/fips_enabled` to `1`. The Go runtime reads this file.
 
-If the Go runtime detects a FIPS preference, it configures OpenSSL during program initialization. This includes disabling FIPS mode if `GOFIPS=0` even if OpenSSL is configured to be in FIPS mode by default. If configuration fails, program initialization panics.
+If the Go runtime detects `GODEBUG=fips140=on` and OpenSSL is not using a FIPS-compliant engine or provider, the program will panic during program initialization. This may be useful to detect and refuse to run with incorrectly configured OpenSSL installations. Otherwise, `GODEBUG=fips140` has no effect.
 
-If no preference is detected, the Go runtime doesn't set the OpenSSL FIPS mode, and the standard OpenSSL configuration is left unchanged. For more information about the standard OpenSSL FIPS behavior, see https://www.openssl.org/docs/fips.html.
+For more information about the standard OpenSSL FIPS behavior, see https://www.openssl.org/docs/fips.html.
+
+> [!WARNING]
+> Previous to Go 1.24, setting `GOFIPS` would make the Go runtime to try to modify the configured FIPS mode. This is no longer supported.
 
 ### Windows FIPS mode (CNG)
 
 To enable FIPS mode on Windows, [enable the Windows FIPS policy](https://docs.microsoft.com/en-us/windows/security/threat-protection/fips-140-validation#step-3-enable-the-fips-security-policy).
 
-If the Go runtime detects `GOFIPS=1` and FIPS policy is not enabled, the program will panic during program initialization. This may be useful to detect and refuse to run on incorrectly configured Windows systems. Otherwise, `GOFIPS` has no effect.
-
-> [!NOTE]
-> Unlike `opensslcrypto`, a Windows program built with `cngcrypto` doesn't include the ability to enable/disable FIPS mode. The change must be made by configuring the OS, not the Go program.
->
-> This is because Windows FIPS mode is not a per-process setting, and changing it may require elevated permissions. We expect that adding a feature that attempts to change the Windows policy would have unintended consequences.
+If the Go runtime detects `GODEBUG=fips140=on` and FIPS policy is not enabled, the program will panic during program initialization. This may be useful to detect and refuse to run on incorrectly configured Windows systems. Otherwise, `GODEBUG=fips140` has no effect.
 
 For testing purposes, Windows FIPS policy can be enabled via the registry key `HKLM\SYSTEM\CurrentControlSet\Control\Lsa\FipsAlgorithmPolicy`, dword value `Enabled` set to `1`.
 
 ### macOS FIPS mode (CommonCrypto/CryptoKit)
 
-CommonCrypo/CrytoKit is FIPS compliant by default. This means that regardless of which mode you set `GOFIPS` to, the cryptographic functions will always be FIPS-enabled.
+CommonCrypo/CrytoKit is FIPS compliant by default. This means that regardless of which mode you set `GODEBUG=fips140` or `GOFIPS` to, the cryptographic functions will always be FIPS-enabled.
```

Comment:
It seems like this means this backend will always put the Go runtime into FIPS-enabled mode, considering the ADR. I'm not sure if the implications are interesting enough to be worth mentioning, though.

Reply:
macOS doesn't have the concept of FIPS mode. It being FIPS compliant is an internal detail that is not surfaced in any API. This means that we don't know if the user intent is to really be in FIPS mode or not. We shouldn't set the Go runtime into FIPS-enabled mode (which means `fips140.Enabled` return true and `crypto/tls` only accepts FIPS compliant settings) unless we have a clear signal, and that signal is `GODEBUG=fips140` or `GOFIPS` on macOS, nothing else.

Reply:
I see, that makes sense, but certainly isn't what I'd expect based on the ADR, and I think this text is misleading... `CommonCrypo/CrytoKit is FIPS compliant by default`, and `the cryptographic functions will always be FIPS-enabled`, but `crypto/tls` will accept non-FIPS-compliant settings by default? I don't think the doc establishes the different meanings of FIPS modes/compliance well enough to try to make detailed distinctions like this without more direct explanation.

Reply:
FWIW, the impression I get from this (and maybe the way to document it?) is that the lack of user intent is a problem that causes a potential compatibility issue, and because we prefer compatibility, we choose not to enable Go runtime FIPS mode in this case.

Reply:
> I see, that makes sense, but certainly isn't what I'd expect based on the ADR

Let's discuss this in the next team meeting. It might deserve a new ADR or amending the one we already have.

> I think this text is misleading...

Agree, I've rewritten the macOS section to make it less confusing.

>  I don't think the doc establishes the different meanings of FIPS modes/compliance well enough to try to make detailed distinctions like this without more direct explanation.

Good point. I'll create a follow up PR clearly stating what Go FIPS mode means, and how it might differ from system FIPS mode.